cmd/github.go

```diff
@@ -0,0 +1,351 @@
+	"fmt"
+	"math/rand"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/juju/ansiterm"
+	"github.com/skratchdot/open-golang/open"
+	"github.com/urfave/cli"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/github"
+
+	"github.com/manifoldco/manifold-cli/analytics"
+	"github.com/manifoldco/manifold-cli/api"
+	"github.com/manifoldco/manifold-cli/color"
+	"github.com/manifoldco/manifold-cli/config"
+	"github.com/manifoldco/manifold-cli/generated/identity/client/authentication"
+	"github.com/manifoldco/manifold-cli/generated/identity/client/user"
+	"github.com/manifoldco/manifold-cli/generated/identity/models"
+	"github.com/manifoldco/manifold-cli/prompts"
+)
+
+var (
+	sourceGitHub      = "github"
+	sourceGitHubToken = "github-token"
+	stateChars        = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ123456789"
+)
+
+// githubAuthorizationRequest is the request sent to GitHub for Personal Authorization Tokens
+type githubAuthorizationRequest struct {
+	Scopes      []string `json:"scopes"`
+	Notes       string   `json:"note"`
+	Fingerprint string   `json:"fingerprint"`
+}
+
+// githubAuthorizationResponse is the response that is given for Personal Authorization Token
+// requests
+type githubAuthorizationResponse struct {
+	ID    int
+	Token string
+
+	Errors []struct {
+		Code string
+	}
+}
+
+// githubErrors are the human readable error messages from GitHub during the Personal Authorization
+// Token flow
+var githubErrors = map[string]string{
+	"already_exists": "A Personal Authorization Token already exists for Manifold",
+}
+
+// oauthStoreFunc is responsible for storing the OAuth authentication with Manifold, either by
+// linking accounts or logging in
+type oauthStoreFunc func(ctx context.Context, cfg *config.Config, a *analytics.Analytics, req models.OAuthAuthenticationRequest) error
+
+// githubWithCallback identified with GitHub via. the normal browser-based OAuth flow
+func githubWithCallback(ctx context.Context, cfg *config.Config, a *analytics.Analytics, store oauthStoreFunc) error {
+	authConfig := &oauth2.Config{
+		ClientID:    config.GitHubClientID,
+		Scopes:      []string{"user"},
+		Endpoint:    github.Endpoint,
+		RedirectURL: "http://127.0.0.1:49152/github/callback",
+	}
+
+	// set up the oauth client
+	state := genRandomString(12)
+	url := authConfig.AuthCodeURL(state, oauth2.AccessTypeOffline)
+
+	// let the user auth in the browser
+	authMsg := color.Color(ansiterm.Blue,
+		"A browser window will open for authentication. Please close this browser when finished")
+	fmt.Println(authMsg)
+
+	time.Sleep(time.Second * 2) // give them time to read the message
+
+	err := open.Start(url)
+	if err != nil {
+		return cli.NewExitError("Unable to open authorization in browser", -1)
+	}
+
+	// handle http callback
+	done := make(chan bool)
+	errCh := make(chan error)
+	go func() {
+		http.HandleFunc("/github/callback", githubCallback(ctx, cfg, state, store, a, done, errCh))
+
+		errCh <- http.ListenAndServe(":49152", nil)
```

Comment:
Should we make this actually look up a free/available port? on top of also be configurable?

Reply:
Something like ```port, err := net.Listen("tcp", ":0")```

Reply:
Yeah, the tricky part is it has to match the redirect URI on GitHub, or it'll come back with an error on the login saying the redirect URIs didn't match